File: calims2-api/src/java/gov/nih/nci/calims2/business/inventory/container/CoordinateHelper.java
/*L
 * Copyright Moxie Informatics.
 *
 * Distributed under the OSI-approved BSD 3-Clause License.
 * See http://ncip.github.com/calims/LICENSE.txt for details.
 */

/**
 *
 */
package gov.nih.nci.calims2.business.inventory.container;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang.StringUtils;

import gov.nih.nci.calims2.domain.inventory.Coordinate;
import gov.nih.nci.calims2.domain.inventory.Layout;
import gov.nih.nci.calims2.domain.inventory.enumeration.LayoutLabelType;

/**
 * @author viseem
 * 
 */
public class CoordinateHelper {

  private static final int DIGIT_BASE = 10;
  private static final int LETTER_BASE = 26;

  /**
   * Converts a given integer into a coordinate label according to the label type.
   * 
   * @param labelType The label type
   * @param maximum The maximum number of values in the dimension
   * @param value The value to convert
   * @return The label corresponding to the given value
   */
  public static String getCoordinateString(LayoutLabelType labelType, int maximum, int value) {
    switch (labelType) {
      case DIGITS: {
        String result = Integer.toString(value + 1);
        while (result.length() < 2) {
          result = "0" + result;
        }
        return result;
      }
      case ALL_LOWER_CASE: {
        return convertValue('a', maximum, value);
      }
      case ALL_UPPER_CASE: {
        return convertValue('A', maximum, value);
      }
      default: {
        throw new InternalError("Can not happen");
      }
    }
  }

  private static String convertValue(char firstCharacter, int maximum, int value) {
    int length = (int) Math.floor(Math.log(maximum) / Math.log(LETTER_BASE)) + 1;
    if (value == 0) {
      return StringUtils.repeat(Character.toString(firstCharacter), length);
    }
    String result = "";
    int remaining = value;
    while (remaining > 0) {
      result = Character.toString((char) (firstCharacter + remaining % LETTER_BASE)) + result;
      remaining /= LETTER_BASE;
    }
    while (result.length() < length) {
      result = firstCharacter + result;
    }
    return result;
  }

  /**
   * Converts a given String into the integer coordinate according to the label type.
   * 
   * @param labelType The label type
   * @param value The string to convert
   * @return The coordinate corresponding to the given label.
   */
  public static int getCoordinateValue(LayoutLabelType labelType, String value) {
    if (StringUtils.isBlank(value)) {
      return 0;
    }
    switch (labelType) {
      case DIGITS: {
        return Integer.parseInt(value) - 1;
      }
      case ALL_LOWER_CASE: {
        return convertString('a', value);
      }
      case ALL_UPPER_CASE: {
        return convertString('A', value);
      }
      default: {
        throw new InternalError("Can not happen");
      }
    }
  }

  private static int convertString(char firstCharacter, String value) {
    int result = 0;
    for (int i = 0; i < value.length(); i++) {
      result = result * LETTER_BASE + value.charAt(i) - firstCharacter;
    }
    return result;
  }

  /**
   * Converts a given coordinate in the corresponding coordinate label string.
   * 
   * @param layout The layout
   * @param coordinate The coordinate to convert
   * @return The coordinate label string
   */
  public static String getCoordinateLabel(Layout layout, Coordinate coordinate) {
    String x = getCoordinateString(layout.getX_dimensionLabel(), layout.getX_dimension(), coordinate.getX());
    String y = getCoordinateString(layout.getY_dimensionLabel(), layout.getY_dimension(), coordinate.getY());
    return x + ", " + y;
  }

  /**
   * Converts a given coordinate in the corresponding coordinate suffix string.
   * 
   * @param layout The layout
   * @param coordinate The coordinate to convert
   * @return The coordinate suffix string.
   */
  public static String getCoordinateSuffix(Layout layout, Coordinate coordinate) {
    String x = getCoordinateString(layout.getX_dimensionLabel(), layout.getX_dimension(), coordinate.getX());
    String y = getCoordinateString(layout.getY_dimensionLabel(), layout.getY_dimension(), coordinate.getY());
    return "_" + x + y;
  }

  /**
   * Converts a given list of coordinates in the corresponding list of coordinate label string.
   * 
   * @param layout The layout
   * @param coordinates The coordinates to convert
   * @return The converted coordinates
   */
  public static List<String> getCoordinateLabels(Layout layout, List<Coordinate> coordinates) {
    List<String> labels = new ArrayList<String>();
    for (Coordinate coordinate : coordinates) {
      labels.add(getCoordinateLabel(layout, coordinate));
    }
    return labels;
  }

  /**
   * Calculates the maximum possible length of a suffix for the given layout.
   * 
   * @param layout The layout
   * @return The maximum possible length of a suffix for the given layout.
   */
  public static int getMaximumSuffixLength(Layout layout) {
    return 1 + getMaximumLength(layout.getX_dimensionLabel(), layout.getX_dimension())
        + getMaximumLength(layout.getY_dimensionLabel(), layout.getY_dimension());
  }

  private static int getMaximumLength(LayoutLabelType labelType, int maximum) {
    if (labelType == LayoutLabelType.DIGITS) {
      int length = (int) Math.floor(Math.log(maximum) / Math.log(DIGIT_BASE)) + 1;
      return (length > 1) ? length : 2;
    }
    return (int) Math.floor(Math.log(maximum) / Math.log(LETTER_BASE)) + 1;
  }
}

File: calims2-webapp/src/dojo/modules/calims2/inventory/layout/CoordinateHelper.js
dojo.provide("calims2.inventory.layout.CoordinateHelper");

dojo.declare(
  "calims2.inventory.layout.CoordinateHelper",
  null,
  {
    // Pattern map for the various label types specified in word
    // This needs to be updated in synch with the LayoutLabelType enumeration 
    patterns : { "ALL_LOWER_CASE" : "^[a-z]*$" , "ALL_UPPER_CASE" : "^[A-Z]*$" , "DIGITS" : "^[0-9]*$"},
    // Converts a given label into the corresponding integer coordinate according to the label type. 
    // The given label is assumed to be valid for the label type
    // labelType The type of label
    // value the value to convert
    // returns the integer coordinate corresponding to the given value
    getCoordinateValue : function(labelType, value) {
      var regexp = new RegExp("^\S*$");
      if (regexp.test(value)) {
        return 0;
      }
      switch (labelType) {
        case "DIGITS": {
          return parseInt(value, 10) - 1;
        }
        case "ALL_LOWER_CASE": {
          return this.convertString("a", value);
        }
        case "ALL_UPPER_CASE": {
          return this.convertString("A", value);
        }
      }
    },
    
    // Converts a given label into the corresponding integer coordinate for the alphabetic label types. 
    // firstCharacter the first character in the range
    // value the value to convert
    // returns the integer coordinate corresponding to the given value
    convertString : function(firstCharacter, value) {
      var firstCode = firstCharacter.charCodeAt(0);
      var result = 0;
      for (var i = 0; i < value.length; i++) {
        result = result * 26 + value.charCodeAt(i) - firstCode;
      }
      return result;
    },
    
    // Validates a given label
    // labelType The type of label
    // maximum The maximum number of values in the dimension
    // The label to validate
    // True if the given label is valid according to the label type and maximum.
    validateCoordinate : function(labelType, maximum, value) {
      var regexp = new RegExp(this.patterns[labelType]);
      if (regexp.test(value)) {
        var coordinate = this.getCoordinateValue(labelType, value);
        if (coordinate < maximum) {
          return true;
        }
      }
      return false;
    },
    
    // Convert the given coordinate into the corresponding label according to the label type and maximum 
    // labelType The type of label
    // maximum The maximum number of values in the dimension
    // value The coordinate to convert to a label
    getCoordinateString : function(labelType, maximum, value) {
      switch (labelType) {
        case "DIGITS": {
          var result = "" + (value + 1);
          while (result.length < 2) {
            result = "0" + result;
          }
          return result;
        }
        case "ALL_LOWER_CASE": {
          return this.convertValue("a", maximum, value);
        }
        case "ALL_UPPER_CASE": {
          return this.convertValue("A", maximum, value);
        }
      }
    },
    
    // firstCharacter the first character in the range
    // maximum The maximum number of values in the dimension
    // value the value to convert
    convertValue : function(firstCharacter, maximum, value) {
      var length = Math.floor(Math.log(maximum) / Math.log(26)) + 1;
      var result = "";
      if (value == 0) {
        for (var i = 0; i < length; i++) {
          result += firstCharacter;
        }  
        return result;
      }
      var firstCharacterCode = firstCharacter.charCodeAt(0);
      var remaining = value;
      while (remaining > 0) {
        result = String.fromCharCode(firstCharacterCode + remaining % 26) + result;
        remaining = Math.floor(remaining/26);
      }
      while (result.length < length) {
        result = firstCharacter + result;
      }
      return result;
    },
    // Gets the maximum possible suffix length for a layout
    getMaximumSuffixLength : function(layout) {
      return 1 + this.getMaximumLength(layout.labelX, parseInt(layout.x, 10)) + 
                 this.getMaximumLength(layout.labelY, parseInt(layout.y, 10));
    },
    // Gets the maximum length of a label
    // labelType The type of label
    // maximum The maximum number of values in the dimension
    getMaximumLength : function(labelType, maximum) {
      if (labelType == "DIGITS") {
        var length = Math.floor(Math.log(maximum) / Math.log(10)) + 1;
        return (length > 1) ? length : 2;
      }
      return Math.floor(Math.log(maximum) / Math.log(26)) + 1;
    }  
  }
  );